rust/pattern/schema/rule.rs

```diff
@@ -126,37 +125,33 @@ fn contains_negations<'a>(mut patterns: impl Iterator<Item = &'a Pattern>) -> bo
     })
 }
 
-fn expect_infer_single_edge(then: &ThingVariable, rule_label: &Label) -> Result<()> {
+fn expect_valid_inference(then: &ThingVariable, rule_label: &Label) -> Result<()> {
     if then.has.len() == 1
         && (then.iid.is_none() && then.isa.is_none() && then.value.is_none() && then.relation.is_none())
         || then.relation.is_some()
             && then.isa.is_some()
             && (then.iid.is_none() && then.has.is_empty() && then.value.is_none())
     {
```

Comment:
I'd suggest splitting this condition into two branches, since each side of the `||` corresponds to an inner if-else branch. That should let us get rid of the `unreachable!()` altogether.

Reply:
It will also read better if you factor the conditions out into their own functions, come to think.